File: Src/recursion/graph.cc
#include <iostream>
#include <queue>            // queue in BFS
#include <unordered_map>    // unordered_map
#include <algorithm>        // find
#include "stdio.h"

struct GraphNode{
    int id;
    std::vector<int> adjacent;
    friend bool operator==(const GraphNode& g1, const GraphNode& g2){
        return g1.id == g2.id;
    }
    GraphNode(int id=-1){
        this->id = id;
    };
    ~GraphNode(){
    };
};

class Graph{
private:
    bool hasPathDFS(GraphNode* source, GraphNode* destination, std::vector<int> visited)
    {        
        visited.push_back(source->id);         

        std::cout << "Visited " << source->id << std::endl;
        for (std::vector<int>::iterator it = visited.begin() ; it != visited.end(); ++it)
            std::cout << " " << *it << " ";
        std::cout << std::endl;

        if(source == destination)
            return true;
        
        std::vector<int>::iterator bufadj = source->adjacent.begin();

        // std::cout << "Adjacent " << source->id << std::endl;
        // for (std::vector<int>::iterator it = source->adjacent.begin() ; it != source->adjacent.end(); ++it)
        //     std::cout << " " << *it << " ";
        // std::cout << std::endl;

        GraphNode* bufnode;
        while(bufadj != source->adjacent.end())
        {
            bufnode = &hashmap[*bufadj];

            auto is_find = std::find(visited.begin(), visited.end(), bufnode->id);
            if(is_find != visited.end()) 
            {
                bufadj++;
                continue;
            }

            if(hasPathDFS(bufnode, destination, visited)){
                return true;
            };
            bufadj++;
        }
        return false;
    }
    bool hasPathBFS(GraphNode* source, GraphNode* destination)
    {
        std::vector<int> visited;
        visited.push_back(source->id);
        std::queue<int> visit_nodes_next;

        for (std::vector<int>::iterator it = source->adjacent.begin() ; it != source->adjacent.end(); ++it)
            visit_nodes_next.push(*it);

        GraphNode* bufnode;
        int idbuf;
        while (!visit_nodes_next.empty())
        {
            idbuf = visit_nodes_next.front();
            visit_nodes_next.pop();

            visited.push_back(idbuf);
            bufnode = &hashmap[idbuf];
            
            std::cout << "Visited " << idbuf << std::endl;
            for (std::vector<int>::iterator it = visited.begin() ; it != visited.end(); ++it)
                std::cout << " " << *it << " ";
            std::cout << std::endl;

            if(bufnode == destination)
                return true;

            for (std::vector<int>::iterator it = bufnode->adjacent.begin() ; it != bufnode->adjacent.end(); ++it)
                visit_nodes_next.push(*it);
        }
        
        return false;
    }
public:
    std::unordered_map<int, GraphNode> hashmap;
    GraphNode* getNode(int id){
        return &hashmap[id];
    }
    void addNode(int id);
    void addEdge(int source, int destination){
        GraphNode* s = getNode(source);
        GraphNode* d = getNode(destination);
        s->adjacent.push_back(d->id);
        d->adjacent.push_back(s->id);
    }

    bool hasPathDFS(int source, int destination){
        std::cout << "Find DFS " << source << " - " << destination << std::endl;

        GraphNode* s = getNode(source);
        GraphNode* d = getNode(destination);

        // std::cout << "Node " << s->id << std::endl;
        // for (std::vector<int>::iterator it = s->adjacent.begin() ; it != s->adjacent.end(); ++it)
        //     std::cout << " " << *it+1 << " ";
        // std::cout << std::endl;

        // std::cout << "Node " << d->id << std::endl;
        // for (std::vector<int>::iterator it = d->adjacent.begin() ; it != d->adjacent.end(); ++it)
        //     std::cout << " " << *it+1 << " "; 
        // std::cout << std::endl;
                   
        std::vector<int> visited;
        return hasPathDFS(s, d, visited);
    }
    bool hasPathBFS(int source, int destination){
        GraphNode* s = getNode(source);
        GraphNode* d = getNode(destination);
        return hasPathBFS(s, d);
    }

};

size_t num_node = 8;
const size_t num_edge = 10;

void testGraph(){
    int edge[num_edge][2] = {{1, 2}, {1, 4}, {1, 3}, {1, 6}, {2, 3}, {3, 6}, {4, 6}, {5, 6}, {6, 8}, {7, 8}};

    Graph solution = Graph();
    GraphNode* bufGraphNode;
    uint8_t i = 0;
    for(i=0; i<num_node; i++)
    {
        solution.hashmap[i] = GraphNode(i);
    }

    for(i=0; i<num_edge; i++)
    {
        solution.addEdge(edge[i][0]-1, edge[i][1]-1);
    }

    std::cout << "Check Node and Edge" << std::endl;
    // std::vector<int>::iterator it, uint8_t j;
    
    for(i=0; i<num_node; i++)
    {
        bufGraphNode = solution.getNode(i);
        std::cout << " Node " << bufGraphNode->id << std::endl;
        std::cout << "Adjacent \n";

        for (std::vector<int>::iterator it = bufGraphNode->adjacent.begin() ; it != bufGraphNode->adjacent.end(); ++it)
            std::cout << " " << *it << " ";

        // it = bufGraphNode->adjacent.begin();
        // for(j=0; j<bufGraphNode->adjacent.size(); ++j)
        // {
        //     std::cout << " " << *it << " ";
        //     it++;
        // }

        std::cout << "\n";
    }

    int source = 1;
    int destination = 8;

    std::cout << "DFS PATH? " << solution.hasPathDFS(source-1, destination-1) << std::endl << "--------------\n";
    std::cout << "BFS PATH? " << solution.hasPathBFS(source-1, destination-1) << std::endl;
  
}

File: py/recursion/graph.py
class GraphNode(object):
    def __init__(self, node) -> None:
        self.node = node
        self.adjecent = []


class Solution(object):
    def __init__(self):
        self.hashmap = {}


    def add_edge(self, source, destination):
        print("Add edge from {} to {}".format(source, destination))
        source_node : GraphNode = self.hashmap[source]
        destination_node : GraphNode = self.hashmap[destination]
        source_node.adjecent.append(destination_node)
        destination_node.adjecent.append(source_node)

    def _has_path_DFS(self, source_node: GraphNode, destination_node: GraphNode, visited: set):
        if source_node.node in visited: # ???
            return False, visited

        visited.add(source_node.node)
        print(visited, source_node.node, destination_node.node, len(source_node.adjecent), source_node.adjecent[0].node)
        
        if source_node.node == destination_node.node:
            return True, visited

        limit = len(source_node.adjecent)
        idadj = 0
        while (idadj < limit):
            if source_node.adjecent[idadj].node in visited:
                idadj += 1
                continue
            
            result, visited = self._has_path_DFS(source_node.adjecent[idadj], destination_node, visited)
            if result: 
                return True, visited
            idadj += 1
            
        visited.pop()
        return False, visited
        
    def has_path_DFS(self, source, destination):
        src_node = self.hashmap[source]
        dest_node = self.hashmap[destination]
        visited = set()
        return self._has_path_DFS(src_node, dest_node, visited)        


    def _has_path_BFS(self, source_node: GraphNode, destination_node: GraphNode): # TODO
        visited = set()
        next_to_visit = []

        next_to_visit.append(source_node)

        level = 0
        prev_level_node = source_node
        while(len(next_to_visit) > 0):
            node: GraphNode = next_to_visit.pop(0)

            if node.node == destination_node.node:
                visited.add(node.node)
                return True, visited

            if node.node in visited:
                continue

            visited.add(node.node)
            next_to_visit.extend(node.adjecent) 
            if node == prev_level_node:
                prev_level_node = node.adjecent[-1]
                level += 1
                print(f"Level {level} {prev_level_node.node}")
                
        return False, visited


    def has_path_BFS(self, source, destination):
        return self._has_path_BFS(self.hashmap[source], self.hashmap[destination])
    

    def shortest_path_BFS(self, source_node, destination_node): # TODO How to get shortest path, consider weight and if negative path exists
        visited = set()
        next_to_visit = []

        next_to_visit.append(source_node)

        while(len(next_to_visit) > 0):
            node: GraphNode = next_to_visit.pop(0)

            for next_node in node.adjecent:
                if next_node.node not in visited:                    
                    next_to_visit.extend(next_node) 
                    visited.add(next_node.node)
                
        return visited




def test_graph():
    edge = ((1, 2), (1, 4), (1, 3), (1, 6), (2, 3), (3, 6), (4, 6), (5, 6), (6, 8), (7, 8))
    sol = Solution()

    for i in range(1, 9):
        sol.hashmap[i] = GraphNode(i)

    for i in range(10):
        sol.add_edge(edge[i][0], edge[i][1]);

    for node in sol.hashmap.values():
        print(node.node, [_node.node for _node in node.adjecent])

    print("-"*10)

    source = 1;
    destination = 8;
    # 1 2 3 6 8 # TODO DFS Choose this, but it is not the shortest path
    # 1 3 6 8
    print(f"DFS PATH? {sol.has_path_DFS(source, destination)}")
    print(f"BFS PATH? {sol.has_path_BFS(source, destination)}") # TODO BFS has error
